library/src/main/java/com/davidstemmer/screenplay/flow/Screenplay.java

```diff
@@ -14,73 +14,94 @@
 
     private final Stage stage;
 
-    private final ArrayDeque<Scene> previousScenes = new ArrayDeque<Scene>();
+    private final ArrayDeque<Scene> previousBackstack = new ArrayDeque<>();
 
     public Screenplay(Stage stage) {
         this.stage = stage;
     }
 
     @Override
-    public void go(Backstack nextBackstack, Flow.Direction direction, Flow.Callback callback) {
+    public void go(Backstack backstack, Flow.Direction direction, Flow.Callback callback) {
 
-        Scene incomingScene = (Scene) nextBackstack.current().getScreen();
+        ArrayDeque<Scene> nextBackstack = deque(backstack);
 
-        ArrayDeque<Scene> incomingScenes;
-        ArrayDeque<Scene> outgoingScenes;
+        ArrayDeque<Scene> animatedScenesIn;
+        ArrayDeque<Scene> animatedScenesOut;
 
         SceneCut.Builder sceneCut = new SceneCut.Builder()
                 .setDirection(direction)
                 .setCallback(callback);
 
         if (direction == Flow.Direction.BACKWARD) {
-            Scene outgoingScene = previousScenes.getLast();
-            if (outgoingScene.isStacking()) {
-                incomingScenes = new ArrayDeque<>();
-            } else {
-                incomingScenes = getLastSceneBlock(nextBackstack);
-            }
-            outgoingScenes = new ArrayDeque<>();
-            outgoingScenes.push(outgoingScene);
+            ArrayDeque<Scene> difference = difference(previousBackstack, nextBackstack);
+            animatedScenesOut = getLastSceneBlock(difference);
+            animatedScenesIn = moveToNewSceneBlock(difference) ?
+                    getLastSceneBlock(nextBackstack) :
+                    new ArrayDeque<Scene>();
         }
         else {
-            // Using forward/replace, there is always one (and only one) incoming scene
-            incomingScenes = new ArrayDeque<>();
-            incomingScenes.push(incomingScene);
-            // Forward/replace to a stacked scene:
-            // No outgoing scenes to animate or tear down
-            if (incomingScene.isStacking() || previousScenes.isEmpty()) {
-                outgoingScenes = new ArrayDeque<>();
-            }
-            // Forward/replace to a non-stacked scene:
-            // Animate and tear down the previous scene block
-            else {
-                outgoingScenes = previousScenes;
-            }
+            ArrayDeque<Scene> difference = difference(nextBackstack, previousBackstack);
+
+            animatedScenesIn = getLastSceneBlock(difference);
+            animatedScenesOut = moveToNewSceneBlock(difference) ?
+                    getLastSceneBlock(previousBackstack) :
+                    new ArrayDeque<Scene>();
 
         }
 
         Scene.Transformer delegatedTransformer = direction == Flow.Direction.BACKWARD ?
-                outgoingScenes.getFirst().getTransformer():
-                incomingScene.getTransformer();
+                animatedScenesOut.iterator().next().getTransformer():
```

Comment:
`incomingScenes` or `outgoingScenes` (I changed the name) should only be empty if the destination backstack is the same as the origin backstack. This is actually a problem that I'm seeing in the Flow 0.11 update, where adding the initial scene to the stack calls a `replace` operation with identical origin/destination backstacks. Since I'm going to fix it on that branch I'm going to hold off fixing it here.